tests/e2e/preferential_topology.go

```diff
@@ -0,0 +1,786 @@
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		csiReplicas = *csiDeployment.Spec.Replicas
+
+		// fetching list of datatstores shared between vm's
+		shareddatastoreListMap, err = getListOfSharedDatastoresBetweenVMs(masterIp, dataCenters)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// fetching list of datastores available in different racks
+		rack1DatastoreListMap, err = getListOfDatastoresByClusterName(masterIp, clusters[0])
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		rack2DatastoreListMap, err = getListOfDatastoresByClusterName(masterIp, clusters[1])
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		rack3DatastoreListMap, err = getListOfDatastoresByClusterName(masterIp, clusters[2])
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// fetching list of datastores which is specific to each rack
+		nonShareddatastoreListMapRack1 = getNonSharedDatastoresInCluster(rack1DatastoreListMap,
+			shareddatastoreListMap)
+		nonShareddatastoreListMapRack2 = getNonSharedDatastoresInCluster(rack2DatastoreListMap,
+			shareddatastoreListMap)
+		nonShareddatastoreListMapRack3 = getNonSharedDatastoresInCluster(rack3DatastoreListMap,
+			shareddatastoreListMap)
+
+		// Get different allowed topologies required for creating Storage Class
+		allowedTopologyForRack1 = getTopologySelector(topologyAffinityDetails, topologyCategories,
+			topologyLength, leafNode, leafNodeTag0)
+		allowedTopologyForRack2 = getTopologySelector(topologyAffinityDetails, topologyCategories,
+			topologyLength, leafNode, leafNodeTag1)
+		allowedTopologyForRack3 = getTopologySelector(topologyAffinityDetails, topologyCategories,
+			topologyLength, leafNode, leafNodeTag2)
+		allowedTopologyRacks = nil
+		allowedTopologyForSC := getTopologySelector(topologyAffinityDetails, topologyCategories,
+			topologyLength)
+		for i := 4; i < len(allowedTopologyForSC); i++ {
+			for j := 0; j < len(allowedTopologyForSC[i].Values); j++ {
+				allowedTopologyRacks = append(allowedTopologyRacks, allowedTopologyForSC[i].Values[j])
+			}
+		}
+
+		//set preferred datatsore time interval
+		setPreferredDatastoreTimeInterval(client, ctx, csiNamespace, namespace, csiReplicas)
+	})
+
+	ginkgo.AfterEach(func() {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		ginkgo.By(fmt.Sprintf("Deleting all statefulsets in namespace: %v", namespace))
+		fss.DeleteAllStatefulSets(client, namespace)
+		ginkgo.By(fmt.Sprintf("Deleting service nginx in namespace: %v", namespace))
+		err := client.CoreV1().Services(namespace).Delete(ctx, servicename, *metav1.NewDeleteOptions(0))
+		if !apierrors.IsNotFound(err) {
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}
+		framework.Logf("Perform preferred datastore tags cleanup after test completion")
+		err = deleteTagCreatedForPreferredDatastore(masterIp, allowedTopologyRacks)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		framework.Logf("Recreate preferred datastore tags post cleanup")
```

Comment:
We are only creating tag and not performing any preferred tag attachment here.
In case any testcase fails in between and due to that preferred tags cleanup is not happened, in order it should not affect the other testcases we are performing cleanup here.

And post that we are recreating preferred tags/category for next testcase to consume it. 

Preferred tag attachment we are doing based on the testcase scenarios.